src/scenechange/mod.rs

```diff
@@ -7,69 +7,202 @@
-    self.last_frame = Some((frame_num, ref_frame));
+  /// Determines if the second frame in `frame_subset` should be a keyframe.
+  ///
+  /// The fact that the first frame in `frame_subset` refers to the previous frame
+  /// is an implementation detail and would make the public API more confusing,
+  /// which is why this method should remain private.
+  #[allow(clippy::wrong_self_convention)]
+  fn is_key_frame<T: Pixel>(
+    &mut self,
+    frame_subset: &[Arc<Frame<T>>],
+    frameno: u64,
+    config: &EncoderConfig,
+    inter_cfg: &InterConfig,
+    keyframes: &mut BTreeSet<u64>,
+  ) -> bool {
+    // Find the distance to the previous keyframe.
+    let previous_keyframe = keyframes.iter().last().unwrap();
+    let distance = frameno - previous_keyframe;
+
+    // Handle minimum and maximum key frame intervals.
+    if distance < config.min_key_frame_interval {
+      return false;
+    }
+    if distance >= config.max_key_frame_interval {
+      return true;
+    }
+
+    // Skip smart scene detection if it's disabled
+    if config.speed_settings.no_scene_detection {
+      return false;
+    }
+
+    self.detect_scene_change(frame_subset, frameno, inter_cfg)
   }
 
-  pub fn detect_scene_change(&self, curr_frame: &Frame<T>, frame_num: usize) -> bool {
-    match self.last_frame {
-      Some((last_num, ref last_frame)) if last_num == frame_num - 1 => {
-        let len = curr_frame.planes[0].cfg.width * curr_frame.planes[0].cfg.height;
-        let delta_yuv = last_frame.iter().zip(curr_frame.iter())
-          .map(|(last, cur)| (
-            (i16::cast_from(cur.0) - i16::cast_from(last.0)).abs() as u64,
-            (i16::cast_from(cur.1) - i16::cast_from(last.1)).abs() as u64,
-            (i16::cast_from(cur.2) - i16::cast_from(last.2)).abs() as u64
-          )).fold((0, 0, 0), |(ht, st, vt), (h, s, v)| (ht + h, st + s, vt + v));
-        let delta_yuv = (
-          (delta_yuv.0 / len as u64) as u16,
-          (delta_yuv.1 / len as u64) as u16,
-          (delta_yuv.2 / len as u64) as u16
-        );
-        let delta_avg = ((delta_yuv.0 + delta_yuv.1 + delta_yuv.2) / 3) as u8;
-        delta_avg >= self.threshold
+  /// Runs dynamic scenechange detection on the frames in `frame_subset`.
+  /// Specifically, uses lookahead to exclude any frames that count as a short flash
+  /// of content, then checks the previous and current frame for differences
+  /// to determine if they qualify as a scenechange.
+  fn detect_scene_change<T: Pixel>(
+    &mut self,
+    frame_subset: &[Arc<Frame<T>>],
+    frameno: u64,
+    inter_cfg: &InterConfig,
+  ) -> bool {
+    if self.excluded_frames.contains(&frameno) {
+      // This frame has already been excluded as a candidate for keyframe selection.
+      return false;
+    }
+
+    // Lookahead to avoid coding short flashes as scenecuts
+    let lookahead_distance = cmp::min(
+      cmp::max(1, inter_cfg.pyramid_depth as usize) + 1,
```

Comment:
Is `inter_cfg.pyramid_depth` always correct? I seem to remember some funny business related to our pyramid dpeth stuff. I could be mistaken.

Reply:
That's a good question. In the x264 code it's setting the lookahead distance to the max number of b-frames in the pyramid. This seemed to be the equivalent in rav1e, but does it make sense to use something else here?

Reply:
It can only be shorter, not longer. (In particular pyramid_depth never changes, but we sometimes drop frames from the pyramid structure to create shorter subgop lengths)

Reply:
Resolved via IRC